README.md

```diff
@@ -1,83 +1,127 @@
 
-### Run and test your app
+    :jlink-modules ["java.sql"]
+    
+There are many modules distributed with the JDK, you can ask `java` to list them all.
 
-You can use jlink generated JRE to run and test your clojure app, by
-which you can verify functionality of your app under customized JRE.
+    $ java --list-modules
+    
+If you only use modules packaged with your JDK, then you can use all of the regular Leiningen commands without issue. When you compile a build a JAR file with `lein build` or execute it with `lein run`, the middleware will make sure that the modules are on the path.
+    
+### External Modules
 
-```
-lein jlink run
-lein jlink test
-```
+If you are using modules that are not distributed with the JDK, you can download and add them to your project with the `:jlink-modules-paths` keyword in your `project.clj`. For instance, if you downloaded the [JavaFX modules][6] from [Gluon's website][7] (referred to as "jmods"), you would unpack them somewhere on your machine and then add a reference to them like so...
 
-A more flexible way is to create a profile in your project:
+    :jlink-module-paths ["/opt/java/javafx-jmods-14.0.1"]
+    
+If you are running on Windows, you will want to escape the `\` character in your paths with `\\`. For example:
 
-```clj
-:profiles {
-  :jlink {:java-cmd "./target/jlink/bin/java"}
-}
-```
+    :jlink-module-paths ["C:\\Program Files\\Java\\javafx-jmods-14.0.1"]
+    
+External modules _only_ work at compile time, they cannot be passed to your default `java` command at runtime. There are two solutions:
+
++ You could download and install the SDK for the module. This plugin supports this but not all modules provide an SDK.
++ Create a runtime image and use the `java` command from the image to execute your project; this is the recommended solution.
+
+#### Adding a Module SDK
+
+If you happen to have the SDK for a module installed and you don't want to build a custom runtime image, you can add the path to the SDK to your `project.clj` file on with the `:jlink-sdk-paths` key. For example...
+
+    :jlink-sdk-paths ["C:\\Program Files\\Java\\javafx-sdk-14\\lib"]
+    
+That being said, building a custom image might be easier since you don't need to install anything except the modules.
+    
+### Buidling a Custom Runtime Environment
+
+A custom runtime environment will build a new Java runtime for your project, including _only_ the modules your project uses. If you have external modules but lack access to an SDK, you will _need_ to create a custom runtime in order to actually run your project.
+
+First you need to let the plugin know that you want to use a custom runtime when interacting with your project. Set the `:jlink-custom-jre` key in your `project.clj` to `true`.
 
-### Assemble and package
+    :jlink-jre-image true
```

Comment:
Why we need this explicit flag when `jlink` plugin/middleware is already enabled?

Reply:
When I was working through the Java module stuff, it looked like you could use "internal" modules (modules packaged with your JDK) in your project and, if you used these, you would need the module flags to build and run but you wouldn't need a custom runtime image. I put the flag in for that scenario.

It's hard to tell how often it would be used in real life and I could be wrong, I haven't put together a project to test this out. Java modules in general seem to be kind of unpopular.

Reply:
hmmm... I suggest to remove this flag and use lein profile to turn on/off jlink

Reply:
Okay, I took out the flag. I didn't really want to test it out anyway.